File: examples/python/custom_tool.py
"""Example: Register a custom tool with the agent."""

import asyncio
import os

from agentlib import Agent, AgentConfig, Tool, ToolContext, ToolResult


class CalculatorTool(Tool):
    """A custom tool that performs basic arithmetic."""

    @property
    def name(self) -> str:
        return "calculator"

    @property
    def description(self) -> str:
        return "Perform basic arithmetic: add, subtract, multiply, divide."

    @property
    def input_schema(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "operation": {
                    "type": "string",
                    "enum": ["add", "subtract", "multiply", "divide"],
                },
                "a": {"type": "number"},
                "b": {"type": "number"},
            },
            "required": ["operation", "a", "b"],
        }

    @property
    def is_read_only(self) -> bool:
        return True

    async def call(self, input: dict, context: ToolContext) -> ToolResult:
        operation = input.get("operation")
        a = input.get("a", 0)
        b = input.get("b", 0)

        try:
            if operation == "add":
                result = a + b
            elif operation == "subtract":
                result = a - b
            elif operation == "multiply":
                result = a * b
            elif operation == "divide":
                if b == 0:
                    return ToolResult.error("Cannot divide by zero")
                result = a / b
            else:
                return ToolResult.error(f"Unknown operation: {operation}")

            return ToolResult.success(str(result))
        except Exception as e:
            return ToolResult.error(str(e))


async def main():
    config = AgentConfig(
        api_key=os.environ.get("ANTHROPIC_AUTH_TOKEN", ""),
        model=os.environ.get("ANTHROPIC_MODEL", "claude-sonnet-4-6"),
    )

    agent = Agent(config)
    agent.register_tool(CalculatorTool())

    async for event in agent.run("What is 123 multiplied by 456?"):
        if event.type.value == "message_delta":
            print(event.data["text"], end="", flush=True)
        elif event.type.value == "complete":
            print(f"\n\n[Complete] {event.data['final_content']}")
        elif event.type.value == "error":
            print(f"\n[Error] {event.data['message']}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/go/custom_tool.go
// Example: Register a custom tool with the agent.
package main

import (
	"context"
	"fmt"
	"os"
	"strconv"

	"github.com/steven0lisa/agent-libs/go"
)

// CalculatorTool performs basic arithmetic.
type CalculatorTool struct{}

func (c *CalculatorTool) Name() string        { return "calculator" }
func (c *CalculatorTool) Description() string { return "Perform basic arithmetic." }
func (c *CalculatorTool) IsReadOnly() bool    { return true }

func (c *CalculatorTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"operation": map[string]any{"type": "string", "enum": []string{"add", "subtract", "multiply", "divide"}},
			"a":         map[string]any{"type": "number"},
			"b":         map[string]any{"type": "number"},
		},
		"required": []string{"operation", "a", "b"},
	}
}

func (c *CalculatorTool) Call(_ context.Context, input map[string]any, _ agentlib.ToolContext) (agentlib.ToolResult, error) {
	op := input["operation"].(string)
	a, _ := strconv.ParseFloat(fmt.Sprintf("%v", input["a"]), 64)
	b, _ := strconv.ParseFloat(fmt.Sprintf("%v", input["b"]), 64)

	var result float64
	switch op {
	case "add":
		result = a + b
	case "subtract":
		result = a - b
	case "multiply":
		result = a * b
	case "divide":
		if b == 0 {
			return agentlib.ToolResult{Content: "Cannot divide by zero", IsError: true}, nil
		}
		result = a / b
	default:
		return agentlib.ToolResult{Content: "Unknown operation: " + op, IsError: true}, nil
	}

	return agentlib.ToolResult{Content: fmt.Sprintf("%g", result), IsError: false}, nil
}

func main() {
	cfg := agentlib.DefaultConfig()
	cfg.APIKey = os.Getenv("ANTHROPIC_AUTH_TOKEN")
	agent := agentlib.NewAgent(cfg)
	agent.RegisterTool(&CalculatorTool{})

	ctx := context.Background()
	for event := range agent.Run(ctx, "What is 123 multiplied by 456?") {
		switch event.Type {
		case agentlib.EventMessageDelta:
			fmt.Print(event.Data["text"])
		case agentlib.EventComplete:
			fmt.Printf("\n\n[Complete] %s\n", event.Data["final_content"])
		case agentlib.EventError:
			fmt.Printf("\n[Error] %s\n", event.Data["message"])
		}
	}
}
